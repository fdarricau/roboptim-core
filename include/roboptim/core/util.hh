// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_UTIL_HH
# define ROBOPTIM_CORE_UTIL_HH
# include <roboptim/core/fwd.hh>
# include <roboptim/core/portability.hh>

# include <vector>
# include <utility>
# include <map>

# include <roboptim/core/twice-differentiable-function.hh>

namespace roboptim
{
  namespace detail
  {
    /// \internal
    /// \brief Copy the content of an Eigen vector into a C array.
    ROBOPTIM_DLLAPI void vector_to_array
    (Function::value_type* dst,
     Function::const_vector_ref src);

    /// \internal
    /// \brief Copy the content of a C array into an Eigen vector.
    ROBOPTIM_DLLAPI void array_to_vector (Function::vector_ref dst,
					  const Function::value_type* src);

    /// \internal
    /// Merge gradients from several functions (each gradient is a line).
    /// The first line of the jacobian is the only one used.
    template <typename T>
    void
    jacobian_from_gradients (DifferentiableFunction::matrix_ref jac,
                             const std::vector<const T*>& c,
                             DifferentiableFunction::const_vector_ref x);
  } // end of namespace detail.

  /// \brief Display a vector.
  template <typename T>
  std::ostream& operator<< (std::ostream&, const std::vector<T>&);

  /// \brief Display a pair.
  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const std::pair<T1, T2>&);

  /// \brief Display a map.
  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream&, const std::map<T1, T2>&);

  /// \brief Display an Eigen object with the appropriate IOFormat.
  template <typename T>
  std::ostream& operator<< (std::ostream&, const Eigen::MatrixBase<T>&);

  /// \brief Convert a sparse matrix into a dense matrix.
  ROBOPTIM_DLLAPI
  GenericFunctionTraits<EigenMatrixDense>::matrix_t sparse_to_dense
  (GenericFunctionTraits<EigenMatrixSparse>::const_matrix_ref m);

  /// \brief Convert a sparse vector into a dense vector.
  ROBOPTIM_DLLAPI
  GenericFunctionTraits<EigenMatrixDense>::vector_t sparse_to_dense
  (GenericFunctionTraits<EigenMatrixSparse>::const_gradient_ref v);

  /// \brief Compare sparse vectors (matrices) using both relative and absolute
  /// tolerances.
  /// \see http://stackoverflow.com/a/15052131/1043187
  ROBOPTIM_DLLAPI
  bool allclose
  (const Eigen::SparseMatrix<double>& a,
   const Eigen::SparseMatrix<double>& b,
   double rtol = Eigen::NumTraits<double>::dummy_precision (),
   double atol = Eigen::NumTraits<double>::epsilon ());

  /// \brief Compare dense vectors (matrices) using both relative and absolute
  /// tolerances.
  /// \see http://stackoverflow.com/a/15052131/1043187
  ROBOPTIM_DLLAPI
  bool allclose
  (const Eigen::Ref<const Eigen::MatrixXd>& a,
   const Eigen::Ref<const Eigen::MatrixXd>& b,
   double rtol = Eigen::NumTraits<double>::dummy_precision (),
   double atol = Eigen::NumTraits<double>::epsilon ());

  /// \brief Copy a sparse block into a sparse matrix.
  /// \param matrix matrix to fill.
  /// \param block block to copy.
  /// \param startRow start row of the block.
  /// \param startCol start col of the block.
  /// \param compress whether to compress the sparse matrix at the end.
  template <typename U>
  void copySparseBlock
  (U& matrix,
   const U& block,
   Function::size_type startRow, Function::size_type startCol,
   bool compress = false);

  /// \brief Apply normalize to a scalar.
  inline double normalize (double x);

  /// \brief Apply normalize to each element of an Eigen vector.
  template <typename T>
  inline T normalize (const T& x);
} // end of namespace roboptim.

# include <roboptim/core/util.hxx>
#endif //! ROBOPTIM_CORE_UTIL_HH
