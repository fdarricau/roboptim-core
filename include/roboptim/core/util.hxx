// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#ifndef ROBOPTIM_CORE_UTIL_HXX
# define ROBOPTIM_CORE_UTIL_HXX
# include <roboptim/core/differentiable-function.hh>

namespace roboptim
{
  namespace detail
  {
    template <typename T>
    void
    jacobian_from_gradients (DifferentiableFunction::matrix_ref jac,
                             const std::vector<const T*>& c,
                             DifferentiableFunction::const_vector_ref x)
    {
      for (DifferentiableFunction::matrix_t::Index i = 0; i < jac.rows (); ++i)
        {
          DifferentiableFunction::jacobian_t grad = c[i]->jacobian (x);
          for (DifferentiableFunction::matrix_t::Index j = 0;
	       j < jac.cols (); ++j)
            jac (i, j) = grad(0, j);
        }
    }
  } // end of namespace detail.

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const std::vector<T>& vect)
  {
    typedef typename std::vector<T>::const_iterator citer_t;

    if (vect.empty ())
      return o << "Empty vector";

    citer_t it = vect.begin ();
    o << *it;
    ++it;

    for (; it != vect.end (); ++it)
      o << ", " << *it;
    return o;
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::pair<T1, T2>& p)
  {
    return o << "(" << p.first << ", " << p.second << ")";
  }

  template <typename T1, typename T2>
  std::ostream& operator<< (std::ostream& o, const std::map<T1,T2>& m)
  {
    typedef typename std::map<T1,T2>::const_iterator citer_t;

    if (m.empty ())
      return o << "{}";

    citer_t it = m.begin ();
    o << "{" << it->first << ": " << it->second;
    ++it;

    for (; it != m.end (); ++it)
      o << ", " << it->first << ": " << it->second;
    return o << "}";
  }

  template <typename T>
  std::ostream& operator<< (std::ostream& o, const Eigen::MatrixBase<T>& matrix)
  {
    Eigen::IOFormat ioformat (Eigen::StreamPrecision,
                              Eigen::DontAlignCols,
                              ",", ", ", "(", ")", "(", ")");
    ioformat.rowSpacer = "";
    o << "[";

    // Matrix
    if (matrix.cols () == 1 || matrix.rows () == 1)
      {
	// Vector
        ioformat = Eigen::IOFormat (Eigen::StreamPrecision,
                                    Eigen::DontAlignCols,
                                    ",", ",", "", "", "(", ")");
	ioformat.rowSpacer = "";
	o << matrix.size ();
      }
    else
      o << matrix.rows () << "," << matrix.cols ();

    o << "]" << matrix.format (ioformat);
    return o;
  }


  template <typename U>
  void copySparseBlock
  (U& matrix,
   const U& block,
   Function::size_type startRow, Function::size_type startCol,
   bool compress)
  {
    typedef U matrix_t;
    typedef typename U::Index index_t;

    // Make sure that the block fits in the matrix
    assert (startRow + block.rows () <= matrix.rows ());
    assert (startCol + block.cols () <= matrix.cols ());

    // Iterate over outer size
    for (int k = 0; k < block.outerSize (); ++k)
      {
	// Iterator over inner size
	for (typename matrix_t::InnerIterator it (block.derived (), k);
	     it; ++it)
          {
	    matrix.coeffRef (static_cast<index_t> (startRow + it.row ()),
			     static_cast<index_t> (startCol + it.col ()))
	      = it.value ();
          }
      }

    // Compress sparse matrix if asked
    if (compress)
      matrix.makeCompressed ();
  }

  inline double normalize (double x)
  {
      return (std::fabs (x) < 1e-8)? 0:x;
  }

  template <typename T>
  inline T normalize (const T& x)
  {
    return (x.array ().abs () < 1e-8).select (0, x);
  }
} // end of namespace roboptim.

#endif //! ROBOPTIM_CORE_UTIL_HXX
