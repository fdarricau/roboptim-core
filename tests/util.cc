// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#define BOOST_TEST_MODULE util

#include "shared-tests/fixture.hh"

#include <iostream>

#include <roboptim/core/io.hh>
#include <roboptim/core/util.hh>

using namespace roboptim;

typedef Function::value_type value_type;
typedef Function::size_type  size_type;

BOOST_FIXTURE_TEST_SUITE (core, TestSuiteConfiguration)

void test_normalization
(boost::shared_ptr<boost::test_tools::output_test_stream> output)
{
  typedef Function::size_type size_type;

  Function::vector_t v (10);
  for (size_type i = 0; i < v.size (); ++i)
    {
      v[i] = std::pow (10., -i);
    }

  (*output) << v << std::endl
            << normalize (v) << std::endl;

  Function::matrix_t m (4,4);
  for (size_type i = 0; i < m.rows (); ++i)
    for (size_type j = 0; j < m.cols (); ++j)
      {
	m(i,j) = std::pow (10., -(m.cols () * i + j));
      }

  (*output) << m << std::endl
            << normalize (m) << std::endl;

}

BOOST_AUTO_TEST_CASE (util)
{
  boost::shared_ptr<boost::test_tools::output_test_stream>
    output = retrievePattern ("util");

  // Check vector display.
  std::vector<int> v;

  (*output) << v << std::endl;

  v.push_back (1);
  (*output) << v << std::endl;

  // Check pair display.
  std::pair<int, std::string> p (42, "foo");
  (*output) << p << std::endl;

  // Check map display.
  std::map<std::string, int> m;
  (*output) << m << std::endl;
  m["foo"] = 42;
  m["bar"] = 1337;
  (*output) << m << std::endl;

  // Test operations on dense matrices
  Eigen::MatrixXd dense_a (5,5);
  Eigen::MatrixXd dense_b (5,5);

  for (int i = 0; i < dense_a.rows (); ++i)
    for (int j = 0; j < dense_a.cols (); ++j)
      {
        dense_a (i,j) = dense_b (i,j) = (double)(i*j);
      }

  BOOST_CHECK (allclose (dense_a, dense_b));

  // Test operations on sparse matrices
  Eigen::SparseMatrix<double> sparse_a (5,5);
  Eigen::SparseMatrix<double> sparse_b (5,5);

  for (int i = 0; i < sparse_a.rows (); ++i)
    for (int j = 0; j < sparse_a.cols (); ++j)
      {
        sparse_a.insert (i,j) = (double)(i*j);
        sparse_b.insert (i,j) = (double)(i*j);
      }

  BOOST_CHECK (allclose (sparse_a, sparse_b));

  const size_type vec_size = 4;
  Function::vector_t eigen_vec (vec_size);
  std::vector<value_type> stl_vec (vec_size);
  eigen_vec.setRandom ();

  // Copy Eigen vector to C array
  detail::vector_to_array (stl_vec.data (), eigen_vec);

  for (size_type i = 0; i < vec_size; ++i)
    BOOST_CHECK_CLOSE (eigen_vec[i],
                       stl_vec[static_cast<size_t> (i)],
                       1e-6);

  // Copy C array to Eigen vector
  Function::vector_t eigen_vec2 (vec_size);
  detail::array_to_vector (eigen_vec2, stl_vec.data ());
  BOOST_CHECK (allclose (eigen_vec, eigen_vec2));

  // Test sparse block copy
  GenericFunctionTraits<EigenMatrixSparse>::matrix_t sparse_c (5,5);
  GenericFunctionTraits<EigenMatrixSparse>::matrix_t sparse_d (5,5);
  GenericFunctionTraits<EigenMatrixSparse>::matrix_t sparse_block (2,2);
  sparse_block.coeffRef (0, 0) = 42.;
  sparse_block.coeffRef (1, 0) = 2.;
  sparse_block.coeffRef (1, 1) = -1.;
  sparse_c = sparse_a;
  sparse_d = sparse_a;
  BOOST_CHECK (allclose (sparse_c, sparse_d));
  Eigen::SparseMatrix<double>::Index startRow = 2;
  Eigen::SparseMatrix<double>::Index startCol = 1;
  for (Eigen::SparseMatrix<double>::Index i = 0; i < sparse_block.rows (); ++i)
    for (Eigen::SparseMatrix<double>::Index j = 0; j < sparse_block.cols (); ++j)
      sparse_d.coeffRef (startRow + i, startCol + j) = sparse_block.coeffRef (i, j);
  copySparseBlock (sparse_c, sparse_block, startRow, startCol);
  BOOST_CHECK (allclose (sparse_c, sparse_d));
  sparse_c.setZero ();
  sparse_c = sparse_a;
  copySparseBlock (sparse_c, sparse_block, startRow, startCol, true);
  BOOST_CHECK (allclose (sparse_c, sparse_d));

  // Test normalization
  test_normalization (output);

  std::cout << output->str () << std::endl;
  BOOST_CHECK (output->match_pattern ());
}

BOOST_AUTO_TEST_SUITE_END ()
