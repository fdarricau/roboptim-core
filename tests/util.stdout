Empty vector
1
(42, foo)
{}
{bar: 1337, foo: 42}
[10](1,0.1,0.01,0.001,0.0001,1e-05,1e-06,1e-07,1e-08,1e-09)
[10](1,0.1,0.01,0.001,0.0001,1e-05,1e-06,1e-07,1e-08,0)
[4,4]((1,0.1,0.01,0.001), (0.0001,1e-05,1e-06,1e-07), (1e-08,1e-09,1e-10,1e-11), (1e-12,1e-13,1e-14,1e-15))
[4,4]((1,0.1,0.01,0.001), (0.0001,1e-05,1e-06,1e-07), (1e-08,0,0,0), (0,0,0,0))
